### ISSUE TITLE
Akiba's Trip Undead & Undressed (333980)

### ISSUE THREAD

**Issue by ghost** [ValveSoftware/Proton#651]
Akiba crashes on startup, due to MFStartup not returning S_OK. This was fixed with Wine 3.8 in this [commit](https://source.winehq.org/git/wine.git/?a=commit;h=beb3094ca8cdf50b5422e7242cb74354f4e03420).

Original upstream bug report can be found [here](https://bugs.winehq.org/show_bug.cgi?id=43607).

Proton: 3.7-4 Beta
Proton Log: [steam-333980.log](https://github.com/ValveSoftware/Proton/files/2321480/steam-333980.log)

**Comment by ghost** [comment]
I've reported the bug that happens after this, which is an infinite black window, [upstream](https://bugs.winehq.org/show_bug.cgi?id=45710).

**Comment by ghost** [comment]
Updated log file for Proton 3.16-6 Beta:
[steam-333980.log](https://github.com/ValveSoftware/Proton/files/2749732/steam-333980.log)

**Comment by zzhiyi** [comment]
Yes. Missing Media Foundation support.

**Comment by HonkingGoose** [comment]
Does this game have issues with cutscenes/cinematic/video playback? I'm not sure after reading trough this issue. Can somebody clarify?

If the game has issues with playback, you can earn extra brownie points by giving information about the media files the game uses and what codecs it uses. :smile:

**Comment by tom-sn** [comment]
Disabling movie in the launcher will get past the blackscreen and into the title screen.
In game will have missing dialogue voice and bgm.
Sfx sounds like foot steps, combat sound, ui, and gameplay wise seems to be running fine, gamepad also detected properly.

[steam-333980.log](https://github.com/ValveSoftware/Proton/files/3158236/steam-333980.log)
[steam-333980-Movie-Disabled.log](https://github.com/ValveSoftware/Proton/files/3158237/steam-333980-Movie-Disabled.log)

**Comment by HonkingGoose** [comment]
Hi @wah123 thanks for the log.

Maybe this is also relevant to the issues?
```
13182.400:0028:0029:warn:module:load_builtin_dll cannot open .so lib for builtin L"Steam2.dll": /home/tommy/.local/share/Steam/steamapps/common/Proton 4.2/dist/lib/wine/steam2.dll.so: cannot open shared object file: No such file or directory
```

**Comment by ghost** [comment]
Extracted information from the 3 main jp video files.

[Full Video Information](https://gist.github.com/fureloka/6fa697ea4766fd75da9e88d06b48c568)

**Short Summery:**
Format: MPEG-4
Format Profile: Base Media
Codec ID: isom (isom/iso2/avc1/mp41

Video
Format: AVC
Format Profile: Main@L4.1 / Main@L3.1 / High@L3.1
Codec ID: avc1

Audio
Format: AAC LC
Codec ID: mp4a-40-2

**Comment by ukbeast** [comment]
Videos work, but are upside down now.
Also there is no voice.
[steam-333980.log](https://github.com/ValveSoftware/Proton/files/10006836/steam-333980.log)

**Comment by vanphong1310** [comment]
The game run fine with wine-lutris-GE-Proton7-43-x86_64, but it is broken with anything else (lutris-7.2, proton & ge-proton all version).
Tested in Lutris 0.5.11

**Comment by ukbeast** [comment]
[Videos are now the correct way up.](https://github.com/ValveSoftware/Proton/wiki/Changelog#available-in-proton-experimental-as-of-20230831)

Still no bgm.

**Comment by ukbeast** [comment]
Videos are upside down again, with no game sound still.
![akiba](https://github.com/ValveSoftware/Proton/assets/6401168/56f22748-d065-42a3-be07-f41464f29614)
[steam-333980.log](https://github.com/ValveSoftware/Proton/files/13592414/steam-333980.log)

**Comment by alasky17** [comment]
@ukbeast Thank you for the update - it seems like something changed in the game code that caused this regression.  We will look into this :)

**Comment by Logge1002** [comment]
Any news regarding this game. Would love to play this on the deck

**Comment by Logge1002** [comment]
Any news regarding this problem?

**Comment by Logge1002** [comment]
Any news about this? Could someone check it with the new proton versions like experimental or ge 9.13?

**Comment by brokenwingsx321** [comment]
Has anyone found a fix for this?

**Comment by TTSChad** [comment]
Have been trying to diagnose this with ProtonTricks and Proton-GE with no luck, the videos play flipped and upside down with regular proton and GE. 
As for in-game audio? Completely busted. 
Looking at the steam forums for the game indicate this has been an issue on windows as well, the devs stated it was due to an "outdated sound driver" without specifying further. 
I suspect this game will need custom fixes to get working, such as re-encoding the video files and some hand tuning to get the audio working, because in game everything runs and looks fine apart from no sound whatsoever.

**Comment by PEPPEsant01** [comment]
I’ve been trying to get Akiba's Trip: Undead & Undress working properly on my Linux system (Bazzite), but I’m encountering a couple of issues related to the game's video playback and audio.

Video Rendering Issue: Some in-game videos appear upside down or distorted. This only seems to affect video content, not regular gameplay.
Audio Glitches: I am experiencing intermittent audio stuttering or skips during these video sequences.
Interestingly, these problems do not occur when playing the same game on PS Vita, so it seems to be specific to the PC version.

I’ve tested multiple versions of Proton, including both official and third-party builds, but the issue persists. I was hoping that someone might have a working configuration that resolves these video and audio issues.

If anyone has faced similar problems or has any solutions to suggest, I’d greatly appreciate the help. Thank you for your time and support!

**Comment by synthetictruism** [comment]
The video issue appears to be fixed in the latest GE (10-15 for me) but there is no audio in-game. I've tried messing with the xaudio dlls, but with no luck. Is this something that could ever be fixed with proton, or is it something that the original deva would have to patch in the game? I've read somewhere that the audio once worked on an earlier build of the game with an earlier proton?

**Comment by delet-this** [comment]
I debugged the game for a bit and found out that audio playback is purposefully disabled when wine/proton is detected. The detection is basically

```c++
bool is_running_in_wine() {
  auto ntdll = GetModuleHandleA("ntdll.dll");
  return ntdll != NULL && GetProcAddress(ntdll, "wine_get_version") != NULL;
}
```

This detection could be fooled by using the HideWineExports registry key. However, the detection is used in 3 places. I'm pretty sure 2 of them are related to the audio system init and audio playback, but one seems to be related to movie/video playback. 

When I patched this whole detection away I indeed wasn't able to get too far in the startup before crashing, presumably when the intro video starts. When I patched out only the audio disabling related call sites I still got problems when the game was about to start after intro animations (probably during audio system init or when menu bgm playback starts).

I was able to get the audio-enabled game started once, I'm not sure if that was due to the proton ge localfixes I tried (`disable_protonmediaconverter()`, xact, wmp11, qasf, quartz...) or if it was completely luck based. I haven't been able to reproduce it yet. It was still not perfect, menu effects and combat voices/effects were there but rest of the sounds were missing (bgm, dialogue voices).

Unless the devs are going to fix something, I think getting the sounds working is going to involve hiding the wine exports or bytepatching. Patching isn't ideal, but idk how else the devs' video workaround could be retained. Otherwise the video issue would also have to be fixed in proton I think.

**Comment by delet-this** [comment]
I messed with it a bit more and found that the following Proton GE gamefix, which simply disables all the devs' workarounds, results in menu and combat effects working but bgm and dialogue voices missing, and videos (at least the intros) still playing in correct orientation:

```python
"""Game fix for Akiba's Trip: Undead & Undressed"""
from protonfixes import util

def main() -> None:
    # Disable devs' workarounds which disable the audio system and flip videos.
    util.protontricks('hidewineexports=enable')
```

If one wants to try this out, it can be placed at `~/.config/protonfixes/localfixes/333980.py` for native steam or `~/.var/app/com.valvesoftware.Steam/.config/protonfixes/localfixes/333980.py` for flatpak steam when using Proton GE.

Notably this doesn't have the current GE's video flip fix env var, which also enables winegstreamer for media. Gstreamer seems to be problematic when this game plays audio, resulting in a hang. Also, I'm guessing the devs' third wine workaround I mentioned earlier wasn't related to crashing after all, but it instead flips videos. Some old wine/proton version probably played videos upside down and the devs tried working around that? So `hidewineexports=enable` also fixes the flipping on modern proton.

I might later try if I can find out the reason for the missing bgm and dialogue. I tried combinations of usual verbs like xact, wmp9, quartz but they didn't seem to help. If I fail at that I'd probably still contribute this current incomplete fix to GE (umu-protonfixes), as it's still better than no audio. Of course anyone else is free to do that as well.

**Comment by synthetictruism** [comment]
thanks so much for looking into this!! I've added your fix into ~/.config/protonfixes/localfixes/, and I've got the basic sound effects playing (without dialogue, music and environmental as you said). I was getting a lot of hangs running GE10-25 (I'm assuming that's the Gstreamer issue you mentioned?), but I've just set it to GE10-20 and it runs well without hanging/crashes, plus the opening video is the correct orientation.

(edited for clarification)

**Comment by synthetictruism** [comment]
In case this is of any help: I've noticed that the game ofter hangs if I have to travel to a location where a cutscene will automatically start (e.g. returning to MOGRA after the initial arena fight). The loading screen starts, the loading bar gets just to the end before it freezes.

**Comment by delet-this** [comment]
Thanks for testing! If the game didn't crash at cutscenes previously on Proton GE then using gstreamer might be a better bet after all. Also because I noticed when the game tries to stream sound files, it runs into issues with the same `MFCreateSourceReaderFromByteStream` function call with both gstreamer and winedmo, it's just that the issues are different. With gstreamer it hangs during that call waiting for some result infinitely inside wine. With winedmo that function ends up returning E_INVALIDARG (0x80070057) instead of the expected S_OK (0x0), so the game doesn't continue with the playback of sound files streamed from disk. I'm guessing larger sound files like bgm and dialogue files are streamed, while small files like effects are just loaded into memory once, explaining why some of the sounds work.

I experimented by patching the game to skip the playback of streamed sound and indeed the game then starts with gstreamer as well, with the basic sound effects working.

Looks like the next step towards full sound would be to figure out what's the issue with that `MFCreateSourceReaderFromByteStream` call. Haven't had much progress with that yet but I'll try to continue. I noticed there was a message ``warn:mfplat:resolver_get_bytestream_url_hint Unrecognized content type "`\xfd\xcc\x14\xd4\x15\xf4{".`` in the logs during that call, but I'm not sure if that really matters here.

**Comment by delet-this** [comment]
Well that took a while but I think I have identified the issue regarding the rest of the sounds, or at least one of them. The log message above was indeed related.

The game has weird implementations for the `IMFByteStream`'s synchronous `Read` ("one-read-a-time") and asynchronous `BeginRead`/`EndRead` ("overlapping reads") functions. The game's implementation of the byte stream only supports one of those at a time depending on its internal state. When decoding media, Wine's MF implementation tries to use sync api when the game's byte stream happens to be in "async mode", which results in the game trying to call ReadFile synchronously on a file opened in async mode (`FILE_FLAG_OVERLAPPED`), in which case nothing is read and Wine ends up trying to decode uninitialized junk from memory instead of the actual audio. That was the junk in the logs from earlier. The game also returns an invalid 0xFFFFFFFF "bytes read" value when this happens.

TL;DR: I got the background music working (at least in menu and during the first fight) with an experimental hack for Proton GE which uses the async byte stream api in certain situations for this game. I'd expect all sounds in `lang_common/sound_bgm/bgm.stm` to work. The sound files in that container are in WMA/ASF format.

Dialogue voices don't work yet however :(. They are in WAV format, which might be related. I didn't notice any immediate problems in the log, but I'll try to continue debugging.

I suppose I could share my current test build of Proton if anyone is eager to try, but I don't know if sharing binaries is allowed here (ping @kisak-valve ?).

I also have no idea if that hack is something that could stay long term or if there's a better fix. Also don't know if it breaks something else. Might have to consult someone who actually knows about MF. The fix seems much easier to do on the game's side 😩, but I don't know if we'd be able to contact the devs.

**Comment by ukbeast** [comment]
> Replying to https://github.com/ValveSoftware/Proton/issues/651#issuecomment-3564552601

You could send a PR over on GloriousEggroll perhaps?

I think proton-ge would be a good compromise.

Also thank you and everyone else for working on this ❤️

**Comment by delet-this** [comment]
It's probably the best idea for now yeah 👍. Though before that I'd still want to take a shot at getting the WAV dialogue voices working as well, and then I'd want to do some testing to check that the rest of the game isn't completely broken with this.

**Comment by synthetictruism** [comment]
That is amazing work! :)
I've tried to email the developers before, back when the Kati DLC was released to ask if there would ever be a patch, but recieved to response from them... I've posted on the Marvelous/XSeed games Discord as well with no response.
If you would like anyone to test the proton build, I'd be very happy to give it a go!

**Comment by delet-this** [comment]
@synthetictruism Here's a test build if you want to try: https://github.com/delet-this/proton-ge-custom/releases/tag/GE-Proton-akiba-1 It's installed the same way as Proton-GE as described [here](https://github.com/GloriousEggroll/proton-ge-custom#native). The sound effects and bgm work but the dialogue doesn't, at least on my end. If you have the previous fix at `~/.config/protonfixes/localfixes/333980.py` then you should delete or rename it. Or if you want to experiment with your own changes then copy over the new one from `compatibilitytools.d/GE-Proton-akiba-1/protonfixes/gamefixes-steam/333980.py` to use as a base.

I haven't made much progress with the dialogue, I'm thinking I might just have to try getting this merged to GE without it. What I found so far: On windows the game opens the evvoice.stm file when the dialogue starts and closes it when the dialogue stops. On proton however the file is closed almost immediately because `IXAudio2SourceVoice::GetState` is reporting zero buffers queued (the game uses this to find when the playback ends). When I force the game to continue regardless I get like unintelligible, very distorted, possibly slowed down audio when the dialogue is playing. It seems to happen on both microsoft's XAudio2 as well as FAudio, so I'm thinking it could be an issue related to decoding the ADPCM WAV dialogue files in mf/mfplat. I might try to re-encode/transcode the files in evvoice.stm at some point, but before that I'd have to reverse engineer the format of that .stm container thingy.

**Comment by synthetictruism** [comment]
@delet-this that's amazing, thanks! I've quickly tried it out and its working well, but I got to head out to work, so I'll test it out a little more thoroughly later tonight and see how it goes. I'm tempted to see if older versions of the game had the same dialogue .stm files, i.e. from before the dlc release.

**Comment by synthetictruism** [comment]
The game with the new Proton runs well, until it very occasionally hangs, usually in an open area when running around after about 5 seconds. It seems pretty random and I can't tell what event is triggering the hang. Music and background effects are all fine, no timing issues or anything. Dialogue is missing as expected. Cutscenes are fine, moving into a location that have a cutscene upon entering are fine too (either traveling there directly, or via the map).

The STM files are a bit odd - it looks like they are RIFF files. There is a header buried in there with "RIFF・  WAVEfmt 2" with a chunk size of 3576 (the hex after it is F0 0D 00 00, so have I got that right? I'm still learning!). I've tried to open it in an RIFF reader, but I'm getting an error saying it has no RIFF header... 

edit: also, the older version of the game (2016 release) has a smaller dialogue file, I'm assuning the latest (2021 latest version) includes the Kati route dialogue. I've not had a chance to test the older verison with the new Proton...

**Comment by delet-this** [comment]
Thanks again for testing!

> occasionally hangs

Could try [enabling Proton logging](https://github.com/ValveSoftware/Proton/wiki/Proton-FAQ#how-to-enable-proton-logs), and then if it happens again, you could post the log here to see if there's something useful related to the hang. But really I should also try reproducing the hangs when I have got some time.

> STM files

I think it's a proprietary container/archive file format that has bunch of Microsoft ADPCM encoded WAV files embedded in it. The `RIFF .... WAVEfmt` thing is in the beginning of each WAV file. If you search for that, there should be multiple matches. The stm file seems to have its own header at the beginning: it begins with 0xFADEBABE "magic bytes"/signature and then it seemingly has the number of WAV files repeated twice: 0x16EE, or 5870 in decimal. When I look up the WAV file signature pattern (`52 49 46 46 ?? ?? ?? ?? 57 41 56 45`) I get 5870 results which checks out. I'm not really sure about the rest of the stm header, it seems like it maybe has some kind of ID for each file, or maybe it's lengths or byte offsets within the file. I could probably find out more if I took the time to reverse engineer the game's code used for reading these files.

If you want to extract the individual WAV files, you can do that with photorec: https://steamcommunity.com/app/333980/discussions/0/527273452881110952/

>older versions

I quickly tried the 2016 and 2021 versions available in the betas but unfortunately at least the first dialogues didn't play on my end :(.

**Comment by synthetictruism** [comment]
@delet-this thanks for all the info! That make a lot more sense now, thanks!!
I'll generate some logs a bit later when I get some time to play it, and post the errors. :) 

I've seen some chatter about msadp32.dll on later versions of Windows being an issue decoding ADPCM (within WINE). Incidentally, I can't open the WAVs in WMP on Win11, so I've had to use VLC and Adobe Audition. There are some posts suggesting using the version of msadp32.dll that came with WinXP works to some degree. Back in 2023 there was a post on Steam from an xseed rep where she said "The developers looked into this issue and found that it is caused by an old sound driver, but unfortunately, this cannot be addressed due to the software used in the game.", and I was wondering if this might be what she was refering to. I'm assuming that this train of thought won't really go anywhere, and the best fix is likely to be a case of re-encoding the sound and somehow converting back into an STM? 

BTW, thank you so much for all teh work you've put into investigating this!!

**Comment by delet-this** [comment]
I now got the dialogue also working by figuring out the format and writing a script that transcodes the ADPCM WAVs to WMA using ffmpeg and rebuilds the .stm file with them, so I'd guess there's something weird/unusual with the encoding of the original audio. Or at least I don't think ADPCM should normally be broken on Proton. I don't know if we should settle for this workaround for now, or if I should try to find out what's wrong with the decoding of the original audio. I don't really know where I would start with tbh.

**Comment by synthetictruism** [comment]
that's amazing! Just checking, the proton build posted further up isn't meant to enable the dialogue, it just activate the SFX and music files?  
I could try the msadp32.dll from XP and see if it helps?

I switched on Proton logging, this is where it errored/hanged:

2007.176:0138:01ec:trace:seh:NtGetContextThread 0xfffffffe: ebp=04d7ff38 esp=04d7fe1c eip=7ad0e678 cs=0023 ss=002b flags=00000246
2007.176:0138:01ec:trace:seh:NtGetContextThread 0xfffffffe: ds=002b es=002b fs=0063 gs=006b
2007.176:0138:01ec:warn:seh:dispatch_exception backtrace: --- Exception 0xc0000005 at 0xf528cb1b: /usr/lib/pressure-vessel/overrides/lib/i386-linux-gnu/libgcc_s.so.1 + 0x28b1b.
2007.176:0138:01ec:trace:seh:dispatch_exception code=c0000005 (EXCEPTION_ACCESS_VIOLATION) flags=0 addr=F528CB1B
2007.176:0138:01ec:trace:seh:dispatch_exception  info[0]=00000000
2007.176:0138:01ec:trace:seh:dispatch_exception  info[1]=FFFFFFFF
2007.176:0138:01ec:trace:seh:dispatch_exception eip=f528cb1b esp=04b7e898 ebp=04b7e970 eflags=00010247
2007.176:0138:01ec:trace:seh:dispatch_exception eax=00000000 ebx=04b7ea18 ecx=f7e92358 edx=00000021
2007.176:0138:01ec:trace:seh:dispatch_exception esi=04b7fd88 edi=04b7e8d0 cs=0023 ds=002b es=002b fs=0063 gs=006b ss=002b
2007.176:0138:01ec:trace:seh:call_seh_handlers calling handler at 7BF6BF90 code=c0000005 flags=0
wine client error:1ec: write: Bad file descriptor
2007.176:0138:01f0:warn:seh:dispatch_exception backtrace: --- Exception 0xc0000005 at 0xf528cb1b: /usr/lib/pressure-vessel/overrides/lib/i386-linux-gnu/libgcc_s.so.1 + 0x28b1b.
2007.176:0138:01f0:trace:seh:dispatch_exception code=c0000005 (EXCEPTION_ACCESS_VIOLATION) flags=0 addr=F528CB1B
2007.176:0138:01f0:trace:seh:dispatch_exception  info[0]=00000000
2007.176:0138:01f0:trace:seh:dispatch_exception  info[1]=FFFFFFFF
2007.176:0138:01f0:trace:seh:dispatch_exception eip=f528cb1b esp=05e4e898 ebp=05e4e970 eflags=00010247
2007.176:0138:01f0:trace:seh:dispatch_exception eax=00000000 ebx=05e4ea18 ecx=f7e92358 edx=00000021
2007.176:0138:01f0:trace:seh:dispatch_exception esi=05e4fd88 edi=05e4e8d0 cs=0023 ds=002b es=002b fs=0063 gs=006b ss=002b
2007.176:0138:01f0:trace:seh:call_seh_handlers calling handler at 79CE8A80 code=c0000005 flags=0
2007.176:0138:01f0:trace:seh:call_seh_handlers handler at 79CE8A80 returned 1
2007.176:0138:01f0:trace:seh:call_seh_handlers calling handler at 7BA381F0 code=c0000005 flags=0
2007.176:0138:01f0:trace:seh:call_seh_handlers handler at 7BA381F0 returned 1
2007.176:0138:01f0:trace:seh:call_seh_handlers calling handler at 7BF6BF90 code=c0000005 flags=0
wine client error:1f0: write: Bad file descriptor
pid 4508 != 4507, skipping destruction (fork without exec?)

**Comment by delet-this** [comment]
> the proton build posted further up isn't meant to enable the dialogue, it just activate the SFX and music files?

Yeah it doesn't have that yet, but I might later add a script to the build that automatically converts the files on startup. If you are on Steam Deck, could you check if the command `ffmpeg` is available in the terminal? It's what I used for the encoding

> this is where it errored/hanged

Thanks, does it have stuff before these lines that seems useful?

**Comment by synthetictruism** [comment]
Just checked on my SD and yes ffmpeg is available :)

Ive looked back and can't see any significant events in the run-up to the hang, but early on in the log there are some GStreamer warnings - wrong ELF class: ELFCLASS64

**Comment by StephenCWills** [comment]
I used the information in this thread to help me get this game running pretty much perfectly on wine-10.0 (Debian Testing). I managed to get through the beginning sequence of the game up through the first visit to MOGRA. I have right-side-up videos, BGMs, voice lines for dialogue, and no hangs or crashes (yet! 🤞). I adapted the changes I needed to a branch on my fork of Wine so others could use or review it.

https://github.com/StephenCWills/wine/tree/akiba

For posterity, here is a rather long summary of the hardships I endured and the fixes I came up with. Hopefully someone can make use of these notes/changes to develop a more user-friendly set of patches for Wine/Proton/Proton-GE.

* To get all videos and BGMs/sounds/voices working, I had to make sure I had the appropriate **32-bit** GStreamer plugins. At the very least, I needed the MPEG4 AVC/H.264 decoder for videos, the asf decoder for BGMs, and the wav parser with the adpcm decoder for voice lines.

* As was previously mentioned in this thread, I had to apply the changes from the following patch in wine-staging to flip the videos over and enable audio:
  https://github.com/wine-staging/wine-staging/blob/master/patches/ntdll-Hide_Wine_Exports/0001-ntdll-Add-support-for-hiding-wine-version-informatio.patch

  Note that this patch must be activated using `regedit` to create a new string value called `HideWineExports` under the `HKCU\Software\Wine` registry key. Set the value of that registry setting to `y`, `t`, or `1`.

* To fix the crashes due to enabled audio, I applied the read-hack patch previously shared in this thread. Note that this patch uses the `SteamGameId` environment variable to disable the hack for any game other than this one. Since I am not running the game through Steam, I had to run the game with `SteamGameId=333980` to activate the hack.

* Deadlocks are caused by a race condition in the winegstreamer media source. The callback for the `SOURCE_ASYNC_REQUEST_SAMPLE` command calls `wait_on_sample()` after entering the `source->cs` critical section. If this happens before invoking the callback for the `SOURCE_ASYNC_START` command, the stream never starts and the sample request never gets any data. Turns out, the Wine developers already fixed this issue at some point since 10.0 so I just pulled the following commit into Debian's wine-10.0 build:
  https://gitlab.winehq.org/wine/wine/-/commit/609e42e9368f96b913b77548e14c000dd0c54563

  Unfortunately, this fix also breaks the movies as the critical section acted as a sort of makeshift flow control for sample requests, but it is necessary to avoid random deadlocks in loading screens. More on that later.

* GStreamer's WAV parser provides `block_align` information to the adpcm decoder, which is used to determine the size of the frames that adpcmdec will process. The WAV files used by the voice lines have a block alignment of 70 bytes, which decompresses to 256 bytes of raw audio samples. As a result, GStreamer provides audio back to winegstreamer in 256-byte chunks, which are passed directly to the game as 256-byte IMFSample objects.

  The game takes the 256 bytes and sends them straight into XAudio2. Each time it provides a buffer to XAudio2, it will turn around and query `IXAudio2SourceVoice::GetState()` to check if `BuffersQueued` is zero. If it is, the game assumes the audio file is complete and immediately shuts down the source voice. Unfortunately, 256 bytes is far too small, and Wine's XAudio2 implementation almost immediately plays all the samples in the buffer and returns zero for `BuffersQueued` so the game shuts down the source voice prematurely.

  To fix this, I modified winegstreamer's media source to accumulate multiple buffers into the `IMFSample`, specifically for audio streams. If you attempt to accumulate video buffers, you will break the videos.

* This game is really picky about the rate at which it receives video data. When Wine was holding the critical section in `wait_on_sample()`, this prevented both `media_stream_RequestSample()` and `source_async_commands_Invoke()` from processing additional `SOURCE_ASYNC_REQUEST_SAMPLE` commands. As I mentioned before, this acted as a sort of makeshift flow control, where the application would wait to receive the previously requested sample before requesting another. Leaving the critical section causes the game to rapidly request all the video's samples all at once, causing the movies to randomly freeze, stutter, fast-forward, or even crash.

  To fix this, I introduced a second critical section in the media stream that can be used to synchronize only the `media_stream_RequestSample()` and `wait_on_sample()` functions. This restores the makeshift flow control without blocking `source_async_commands_Invoke()` from handling other commands such as `SOURCE_ASYNC_START`.

**Comment by delet-this** [comment]
Looks like these fixes are now in Proton GE 10-32!

**Comment by GloriousEggroll** [comment]
Thanks to @StephenCWills for his work on this! I have added the changes to GE-Proton10-32

**Comment by fycher** [comment]
One FMV bug left in 10-32, 10-33 and 10-34: if you pause game with UI menus, the city screen FMV will get paused until level transition, but sound will continue playing
